### DISCUSSION TITLE
External Viewers with Edge or Chrome: encode problem sending PDF page number

### DISCUSSION THREAD

**Discussion by science2002** [sumatrapdfreader/sumatrapdf#2454] [Q&A] [answered]
First let me congratulate: I really enjoy using your great software.
I saw that it is possible to send PDFs and alike to external viewers. Edge (in Win10 64bit) does have "read aloud" and I tried to pass an open file in SumatraPDF to it, starting from the exact page it was in SumatraPDF.
Like Chrome, also Edge accepts page number by adding at the end of the URL the following parameter "#page=100" (if the page is 100 for instance).
So I put - as suggested in SumatraPDF guide -  the following code with shortcut Alt+F+D in advanced settings:
`ExternalViewers [
	[
		CommandLine = "C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"  -inprivate -profile-directory=Default   "%1"#page=%p
		Name = e&Dgium
	]
]`
**Problem**: when it is sent to Edge the page number parameter "#" is encoded to "%23", and the URL address gives error (PDF not opened).
If I use the command line (cmd.exe) to call Edge and sending the PDF address with the page parameter, or I copy and paste its PDF address directly in Edge, the special character # is not encoded, and it works (PDF opened in the proper page). 

**Question**: I guess then that between SumatraPDF & Win the encoding occures. Is there a way to leave "#" as it is when it is sent to Edge? Or is there another way to send a PDF to Edge with its page number?

Thanks for any tip and help.

**Comment by science2002** [comment, marked as answer]
After testing and testing.... and testing here the answer:
Add `file:///` in the code before `"%1"` and it works.

**Comment by GitHubRulesOK** [comment]
The correct call should not be that different, however I agree that using the following is not obvious as to why
```
ExternalViewers [
	[
		CommandLine = "C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe" -inprivate "file:///%1#page=%p"
		Name = e&Dgium
	]
]
```
Is not

**Comment by science2002** [comment]
@GitHubRulesOK  thank you for the answer. 
I take yours as a "refinement" of mine: the different place of quotes  seems do not make much difference if there are no spaces. At least this is what I understood after hours of  trials and errors in attempting to make page number work!

**Comment by GitHubRulesOK** [comment]
the Quotes are to keep the target as one object thus part cosmetic but ensures a file or server with spaces in the name are treated as one continuous text block.
Your `-inprivate` is also part cosmetic (darker theme) as it does not protect from internal exploits in the file unlike SumatraPdf which currently should not invoke internal actions in a PDF.
`--Profile` is usually "Default" but if you have multiple it is a good idea and another may be depending on experience `--no-first-run`

**Comment by science2002** [comment]
Thanks again. I agree with quotes: usually in fact are necessary either for the PDF file path or for the PDF name itself. Without them, the sending file is truncated at the first space found in the full file path.  For that reason I left them just around "%1".
I use -inprivate - if I understood correctly its use - not for security reasons, but simply because I do not need that each PDF is recorded in Edge history and alike (completation URL, etc.). Fortunately the PDF files are well remembered in SumatraPDF: I use even to allow the reopening of  all latest used PDFs.
A furter refinement of the script would be to include a switch to send "via SumatraPDF" also the shortcut to do something in Edge (or Chrome): in my case the shortcut of reading aloud. It is CTRL-Shift-U. Do you know (or does someone know) if there is a switch that allows sending keystrokes to Edge/Chrome, so to include CTRL-Shift-U in above SumatraPDF external viewers script?

**Reply by GitHubRulesOK** [reply]
Googleium thus edge are command line adverse, (though there are many) their logical approach to printing etc is use their API for landscape or provide human keystrokes.
So one possibility is vbs send keys but they are unreliable for a few reasons, first is focus you need to say that Edge is the target window, second is timing since the file may not have fully opened thus needs delays, and lastly you need the status of content to be un wavering from call to call so the same key string will always perform safe actions and not cause a rogue function.
So using a batch file with sendkeys,exe is only slightly better, beware the heavy website advert insertions, the app itself should be clean and signed but you need to double click the zip to see the password for extraction. The zip file may be reported as poor reputation by Yandex as used by script kiddies but the internal exe from http://cpap.com.br/orlando/SendKeysMore.asp should produce a clean file https://www.virustotal.com/gui/file/07e5577d0ad1706d76d680cbd01c78273ac871c140c4905f1189266232a74e56/community

the bat file you call from SumatraPDF using 
cmd /c readAloud.bat "%f" %p

the batch file then has first line the Edge call but for %f you use the %1 and for %p you use %2
the second line is c:\wherever\sendkeys.exe 3 10 "edge window name" "keystring"
HOWEVER a gotcha is that both SumatraPDF and EDGE may both be using the filename as current window name so in the batch file it may be safer to FIRST LINE make a temporary copy of the pdf say %temp%\target.pdf then pass that known new name to edge with the page number then sendkeys to the temporary named window !
**all of above is untested**
